CORDA-3995 Redeliver external events if number of suspends differs

lankydan opened the pull request:
When retrying a flow, only redeliver external events held in a flow's
pending deduplication handlers if there is a difference in the
`currentState`'s `numberOfCommits` or the `numberOfCommits`
the checkpoint has recorded in the database.

If the checkpoint committed, but the flow retried, then the external
events would have been persisted to the database as part of the same
transaction. Therefore there is no need to replay them, as they have
already been processed as saved as part of the checkpoint.

This change is only relevant when the checkpoint persists, but the flow
still needs to retry after this occurs (within the same
transition/event).

lankydan committed:
CORDA-3995 Redeliver external events in number of suspends differs

When retrying a flow, only redeliver external events held in a flow's
pending deduplication handlers if there is a difference in the
`numberOfSuspends` on the `currentState`'s checkpoint or the checkpoint
in the database.

If the checkpoint committed, but the flow retried, then the external
events would have been persisted to the database as part of the same
transaction. Therefore there is no need to replay them, as they have
already been processed as saved as part of the checkpoint.

This change is only relevant when the checkpoint persists, but the flow
still needs to retry after this occurs (within the same
transition/event).

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

lankydan committed:
CORDA-3995 Redeliver external events in number of commits differs

When retrying a flow, only redeliver external events held in a flow's
pending deduplication handlers if there is a difference in the
`numberOfCommits` on the `currentState`'s checkpoint or the checkpoint
in the database.

If the checkpoint committed, but the flow retried, then the external
events would have been persisted to the database as part of the same
transaction. Therefore there is no need to replay them, as they have
already been processed as saved as part of the checkpoint.

This change is only relevant when the checkpoint persists, but the flow
still needs to retry after this occurs (within the same
transition/event).

lankydan commented:
Run tests: slow

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/ErrorFlowTransition.kt:
```diff
@@ -59,32 +59,41 @@ class ErrorFlowTransition(
-                        pendingDeduplicationHandlers = emptyList(),
-                        isRemoved = true
-                )
+                if (currentState.checkpoint.checkpointState.invocationContext.clientId == null) {
```
`!= null` ?  Another checkpoint is recorded if there's a client ID

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -253,29 +256,30 @@ class TopLevelTransition(
+                    actions += Action.AcknowledgeMessages(pendingDeduplicationHandlers)
+                    actions += Action.RemoveSessionBindings(currentState.checkpoint.checkpointState.sessions.keys)
+
+                    if (currentState.checkpoint.checkpointState.invocationContext.clientId == null) {
```
`!= null` ?

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/Action.kt:
```diff
@@ -149,6 +149,8 @@ sealed class Action {
         override fun toString() = "CommitTransaction"
     }
 
+    data class IncrementNumberOfCommits(val currentState: StateMachineState) : Action()
```
What do things look like if you merged this with `CommitTransaction`?  i.e. passed `StateMachineState` into that.  Would ensure we never forgot to increment and since it's counting commits.... makes sense to make it part of commit.

lankydan committed:
CORDA-3995 Redeliver external events if number of commits differs

When retrying a flow, only redeliver external events held in a flow's
pending deduplication handlers if there is a difference in the
`currentState`'s `numberOfCommits` or the `numberOfCommits`
the checkpoint has recorded in the database.

If the checkpoint committed, but the flow retried, then the external
events would have been persisted to the database as part of the same
transaction. Therefore there is no need to replay them, as they have
already been processed as saved as part of the checkpoint.

This change is only relevant when the checkpoint persists, but the flow
still needs to retry after this occurs (within the same
transition/event).

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -575,10 +578,23 @@ internal class SingleThreadedStateMachineManager(
      * Extract all the incomplete deduplication handlers as well as the [ExternalEvent] and [Event.Pause] events from this flows event queue
```
We should update this comment (as it is no longer true).

rick-r3 replied:
Yes.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineState.kt:
```diff
@@ -129,7 +132,8 @@ data class Checkpoint(
                         listOf(topLevelSubFlow),
-                        numberOfSuspends = 0
+                        numberOfSuspends = 0,
+                        numberOfCommits = 1
```
Why does this start at 1? Might be worth comment here explaining that.

rick-r3 replied:
Not sure.  Need to verify this is correct.  I'd assume it should start off the same as the one that in the `StateMachineState` which seems to start at zero.  So I think this should be zero.  Does that break any of the tests?

williamvigorr3 replied:
Changing the number of commits in the `CheckpointState` to start from 1 breaks `error during transition when checkpoint commits but transient db exception is thrown during flow initialisation will retry and complete successfully`. It causes an extra start flow event.

lankydan replied:
Starts at 1 because it's easier to do that here, instead of in `UnstartedFlowTransition`. Where you have to take another copy of the checkpoint to increase the count.

williamvigorr3 replied:
Okay I have added a comment explaining this.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -189,14 +189,18 @@ class TopLevelTransition(
+            val newCheckpoint = startingState.checkpoint.run {
                 val newCheckpointState = if (checkpointState.invocationContext.arguments!!.isNotEmpty()) {
                     checkpointState.copy(
                         invocationContext = checkpointState.invocationContext.copy(arguments = emptyList()),
```
Could we put the `if` here, rather than having two `copy` that we have to keep in sync ?

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/ErrorFlowTransition.kt:
```diff
@@ -59,32 +59,38 @@ class ErrorFlowTransition(
-                )
+                actions += Action.CreateTransaction
+                actions += removeOrPersistCheckpoint
+                actions += Action.PersistDeduplicationFacts(startingState.pendingDeduplicationHandlers)
```
So this means that the in memory state differs from the state in the DB. I guess that is Okay in this cases as the flow aborts.

williamvigorr3 replied:
Are we just removing the pendingDeduplicationHandlers to save memory?

rick-r3 replied:
Not sure I follow this comment.  We ask them to write to the DB here.  Then we ask them to ack right after commit.

williamvigorr3 replied:
So I thought that the pendingDeduplicationHandlers were stored in the same table as the checkpoints but this is not true. So my comment makes no sense.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -161,6 +163,7 @@ class FlowCreator(
             fiber = flowStateMachineImpl,
             anyCheckpointPersisted = existingCheckpoint != null,
             reloadCheckpointAfterSuspendCount = if (reloadCheckpointAfterSuspend) 0 else null,
+            numberOfCommits = 0,
```
What if `existingCheckpoint != null` ?

rick-r3 replied:
@williamvigorr3 I think this was my main concern.  Should be `existingCheckpoint?.checkpointState?.numberOfCommits ?: 0` should it ?

williamvigorr3 replied:
Yes. I think that makes sense. Otherwise if we pause, we will have problems if something goes wrong later for example.

rick-r3 reviewed (commented):
Just refreshing my memory @williamvigorr3.  It's very close.

williamvigorr3 committed:
Add @Suspendable to a test flow.

I am surprised this worked at all.

williamvigorr3 commented:
Run tests: slow

williamvigorr3 committed:
Fix a few minor things based on review.

williamvigorr3 commented:
Run tests: slow

rick-r3 reviewed (approved):
LGTM
